{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "cb1a04b6-7ad1-40bf-8db9-1d8d39d1e2b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sqlite3\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d26ef832-a895-4023-922d-f06041481077",
   "metadata": {},
   "source": [
    "# 1. Check to make sure what column headers are available in my updated weight db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "40ef9850-24bf-4050-ae30-1c76563c1e10",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# All tables: [('weights',), ('sample_info',), ('extra',), ('construction',)]\n",
      "# --------------------\n",
      "('rs10000169', 'ENSG00000163749', 0.17842051839471315, 'T', 'C', 'chr4:77356246:T:C', 'chr4:76435093:T:C')\n",
      "('rs10000239', 'ENSG00000164129', 0.0006143162235295212, 'T', 'C', 'chr4:164368397:T:C', 'chr4:163447245:T:C')\n",
      "('rs10000282', 'ENSG00000250829', -0.009943939242380807, 'C', 'T', 'chr4:187759002:C:T', 'chr4:186837848:C:T')\n",
      "('rs10000282', 'ENSG00000250971', -0.03620415832189286, 'C', 'T', 'chr4:187759002:C:T', 'chr4:186837848:C:T')\n",
      "# --------------------\n",
      "('rs10000169', 'ENSG00000163749', 0.17842051839471315, 'T', 'C', 'chr4:77356246:T:C', 'chr4:76435093:T:C')\n",
      "# --------------------\n",
      "####################\n",
      "[('rs10000169', 'ENSG00000163749', 0.17842051839471315, 'T', 'C', 'chr4:77356246:T:C', 'chr4:76435093:T:C'), ('rs10000239', 'ENSG00000164129', 0.0006143162235295212, 'T', 'C', 'chr4:164368397:T:C', 'chr4:163447245:T:C'), ('rs10000282', 'ENSG00000250829', -0.009943939242380807, 'C', 'T', 'chr4:187759002:C:T', 'chr4:186837848:C:T'), ('rs10000282', 'ENSG00000250971', -0.03620415832189286, 'C', 'T', 'chr4:187759002:C:T', 'chr4:186837848:C:T')]\n",
      "# --------------------\n",
      "['rsid', 'gene', 'weight', 'ref_allele', 'eff_allele', 'snp_id_GRch37', 'snp_id_GRch38']\n"
     ]
    }
   ],
   "source": [
    "# Load a JTI weights file to check the format\n",
    "jti_weights_path = '/data100t1/home/wanying/BioVU/202505_hypophosphatasia/data/harmonized_predixcan_db'\n",
    "weight_fn = 'JTI_Colon_Transverse.with_snp_id.db'\n",
    "conn = sqlite3.connect(f'{jti_weights_path}/{weight_fn}')\n",
    "# conn = sqlite3.connect('/data100t1/share/predixcan_models/gtex_v8/JTI/JTI_Whole_Blood.db')\n",
    "\n",
    "cur = conn.cursor()\n",
    "\n",
    "cur.execute(\"SELECT name FROM sqlite_master WHERE type='table';\")\n",
    "print('# All tables:', cur.fetchall())\n",
    "print('#', '-'*20)\n",
    "\n",
    "cmd = \"SELECT * FROM weights\"\n",
    "cur.execute(cmd)\n",
    "print(cur.fetchone())\n",
    "print(cur.fetchone())\n",
    "print(cur.fetchone())\n",
    "print(cur.fetchone())\n",
    "print('#', '-'*20)\n",
    "\n",
    "snp = 'chr4:76435093:T:C'\n",
    "cmd = f\"SELECT * FROM weights WHERE snp_id_GRch38 = ?\"\n",
    "res = cur.execute(cmd, (snp,)).fetchone()\n",
    "print(res)\n",
    "print('#', '-'*20)\n",
    "\n",
    "print('#' * 20)\n",
    "snps = ['chr4:76435093:T:C', 'chr4:163447245:T:C', 'chr4:163447245:T:C', 'chr4:186837848:C:T']\n",
    "cmd = f\"SELECT * FROM weights WHERE snp_id_GRch38 IN ({','.join('?'*len(snps))})\"\n",
    "res = cur.execute(cmd, snps).fetchall()\n",
    "print(res)\n",
    "print('#', '-'*20)\n",
    "\n",
    "print([v[1] for v in cur.execute(f\"PRAGMA table_info(weights)\").fetchall()])\n",
    "\n",
    "\n",
    "gene = 'ENSG00000164129'\n",
    "cmd = f\"SELECT * FROM weights WHERE gene='{gene}'\"\n",
    "df_weights = pd.'weights',), ('sample_info',), ('extra',), ('construction',)read_sql_query(sql=cmd, con=conn)\n",
    "# df_weights\n",
    "\n",
    "cmd = f\"SELECT * FROM weights\"\n",
    "df_all_weights = pd.read_sql_query(sql=cmd, con=conn)\n",
    "cur.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a578fbe6-c150-43e2-8f45-0a58204ef18d",
   "metadata": {},
   "source": [
    "# 2. Check the updated .db file\n",
    "Make sure ref and alt alleles are matched in org db and updated SNP ids\n",
    "\n",
    "There are some SNPs with strange ref alleles. But it is not due to my conversion.\n",
    "\n",
    "(See code output below) The original JTI file already has those errors\n",
    "\n",
    "For example:\n",
    "* rs199942225\tENSG00000264404\t0.337823\tTRUE\tT\n",
    "* rs372054324\tENSG00000264404\t0.162650\tTRUE\tT\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "6efdab5f-fff8-49d8-a967-5ea6909eec7c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Brain_Spinal_cord_cervical_c-1\n",
      "# Brain_Hippocampus\n",
      "# Brain_Cortex\n",
      "# Brain_Nucleus_accumbens_basal_ganglia\n",
      "# Brain_Caudate_basal_ganglia\n",
      "# Brain_Substantia_nigra\n",
      "# Brain_Anterior_cingulate_cortex_BA24\n",
      "# Brain_Cerebellar_Hemisphere\n",
      "# - org_ref != ref_b38\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>gene</th>\n",
       "      <th>weight</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>eff_allele</th>\n",
       "      <th>snp_id_GRch37</th>\n",
       "      <th>snp_id_GRch38</th>\n",
       "      <th>ref_b37</th>\n",
       "      <th>alt_b37</th>\n",
       "      <th>ref_b38</th>\n",
       "      <th>alt_b38</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37595</th>\n",
       "      <td>rs199942225</td>\n",
       "      <td>ENSG00000265203</td>\n",
       "      <td>-0.010954</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr10:48953923:TRUE:T</td>\n",
       "      <td>chr10:46785439:TRUE:T</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>T</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37596</th>\n",
       "      <td>rs199942225</td>\n",
       "      <td>ENSG00000265763</td>\n",
       "      <td>-0.171888</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr10:48953923:TRUE:T</td>\n",
       "      <td>chr10:46785439:TRUE:T</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>T</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55425</th>\n",
       "      <td>rs372054324</td>\n",
       "      <td>ENSG00000265203</td>\n",
       "      <td>-0.143357</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr10:48992156:TRUE:T</td>\n",
       "      <td>chr10:46747206:TRUE:T</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>T</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              rsid             gene    weight ref_allele eff_allele  \\\n",
       "37595  rs199942225  ENSG00000265203 -0.010954          C          T   \n",
       "37596  rs199942225  ENSG00000265763 -0.171888          C          T   \n",
       "55425  rs372054324  ENSG00000265203 -0.143357          C          T   \n",
       "\n",
       "               snp_id_GRch37          snp_id_GRch38 ref_b37 alt_b37 ref_b38  \\\n",
       "37595  chr10:48953923:TRUE:T  chr10:46785439:TRUE:T    TRUE       T    TRUE   \n",
       "37596  chr10:48953923:TRUE:T  chr10:46785439:TRUE:T    TRUE       T    TRUE   \n",
       "55425  chr10:48992156:TRUE:T  chr10:46747206:TRUE:T    TRUE       T    TRUE   \n",
       "\n",
       "      alt_b38  \n",
       "37595       T  \n",
       "37596       T  \n",
       "55425       T  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Brain_Frontal_Cortex_BA9\n",
      "# Brain_Cerebellum\n",
      "# - org_ref != ref_b38\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>gene</th>\n",
       "      <th>weight</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>eff_allele</th>\n",
       "      <th>snp_id_GRch37</th>\n",
       "      <th>snp_id_GRch38</th>\n",
       "      <th>ref_b37</th>\n",
       "      <th>alt_b37</th>\n",
       "      <th>ref_b38</th>\n",
       "      <th>alt_b38</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>44627</th>\n",
       "      <td>rs199942225</td>\n",
       "      <td>ENSG00000265203</td>\n",
       "      <td>-0.064616</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr10:48953923:TRUE:T</td>\n",
       "      <td>chr10:46785439:TRUE:T</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>T</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65620</th>\n",
       "      <td>rs372054324</td>\n",
       "      <td>ENSG00000265203</td>\n",
       "      <td>-0.214592</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr10:48992156:TRUE:T</td>\n",
       "      <td>chr10:46747206:TRUE:T</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>T</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              rsid             gene    weight ref_allele eff_allele  \\\n",
       "44627  rs199942225  ENSG00000265203 -0.064616          C          T   \n",
       "65620  rs372054324  ENSG00000265203 -0.214592          C          T   \n",
       "\n",
       "               snp_id_GRch37          snp_id_GRch38 ref_b37 alt_b37 ref_b38  \\\n",
       "44627  chr10:48953923:TRUE:T  chr10:46785439:TRUE:T    TRUE       T    TRUE   \n",
       "65620  chr10:48992156:TRUE:T  chr10:46747206:TRUE:T    TRUE       T    TRUE   \n",
       "\n",
       "      alt_b38  \n",
       "44627       T  \n",
       "65620       T  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Brain_Hypothalamus\n",
      "# Brain_Amygdala\n",
      "# Brain_Putamen_basal_ganglia\n"
     ]
    }
   ],
   "source": [
    "# Load a JTI weights file to check the format\n",
    "jti_weights_path = '/data100t1/home/wanying/BioVU/202505_hypophosphatasia/data/harmonized_predixcan_db'\n",
    "\n",
    "for weight_fn in os.listdir(jti_weights_path):\n",
    "    if weight_fn.endswith('.db'):\n",
    "        # JTI_Colon_Transverse.with_snp_id.db\n",
    "        tissue = weight_fn.split('.')[0].split('JTI_')[-1]\n",
    "        if 'Brain' not in weight_fn:\n",
    "            continue\n",
    "        print('#', tissue)\n",
    "        conn = sqlite3.connect(f'{jti_weights_path}/{weight_fn}')\n",
    "        cur = conn.cursor()\n",
    "        \n",
    "        cmd = f\"SELECT * FROM weights\"\n",
    "        df_all_weights = pd.read_sql_query(sql=cmd, con=conn)\n",
    "        \n",
    "        df_all_weights = df_all_weights.dropna().copy()\n",
    "        df_all_weights['ref_b37'] = df_all_weights['snp_id_GRch37'].apply(lambda x: x.split(':')[-2])\n",
    "        df_all_weights['alt_b37'] = df_all_weights['snp_id_GRch37'].apply(lambda x: x.split(':')[-1])\n",
    "        \n",
    "        df_all_weights['ref_b38'] = df_all_weights['snp_id_GRch38'].apply(lambda x: x.split(':')[-2])\n",
    "        df_all_weights['alt_b38'] = df_all_weights['snp_id_GRch38'].apply(lambda x: x.split(':')[-1])\n",
    "        if len(df_all_weights[df_all_weights['ref_b37']!=df_all_weights['ref_b38']]) !=0:\n",
    "            print('# - B37 ref != B38 ref')\n",
    "        if len(df_all_weights[df_all_weights['alt_b37']!=df_all_weights['alt_b38']]) !=0:\n",
    "            print('# - B37 alt != B38 alt')\n",
    "        if len(df_all_weights[df_all_weights['ref_allele']!=df_all_weights['ref_b38']]) !=0:\n",
    "            print('# - org_ref != ref_b38')\n",
    "            display(df_all_weights[df_all_weights['ref_allele']!=df_all_weights['ref_b38']])\n",
    "        if len(df_all_weights[df_all_weights['eff_allele']!=df_all_weights['alt_b38']]) !=0:\n",
    "            print('# - org_alt != alt_b38')\n",
    "            display(df_all_weights[df_all_weights['eff_allele']!=df_all_weights['alt_b38']])\n",
    "        cur.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "dcc4b75d-d6eb-43c3-9608-5721d9ec99f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>gene</th>\n",
       "      <th>weight</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>eff_allele</th>\n",
       "      <th>snp_id_GRch37</th>\n",
       "      <th>snp_id_GRch38</th>\n",
       "      <th>ref_b37</th>\n",
       "      <th>alt_b37</th>\n",
       "      <th>ref_b38</th>\n",
       "      <th>alt_b38</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [rsid, gene, weight, ref_allele, eff_allele, snp_id_GRch37, snp_id_GRch38, ref_b37, alt_b37, ref_b38, alt_b38]\n",
       "Index: []"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all_weights[df_all_weights['snp_id_GRch37'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "cf807280-3e03-41d2-8d10-215b7af593b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>gene</th>\n",
       "      <th>weight</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>eff_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>100136</th>\n",
       "      <td>rs199942225</td>\n",
       "      <td>ENSG00000264404</td>\n",
       "      <td>0.337823</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100137</th>\n",
       "      <td>rs372054324</td>\n",
       "      <td>ENSG00000264404</td>\n",
       "      <td>0.162650</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               rsid             gene    weight ref_allele eff_allele\n",
       "100136  rs199942225  ENSG00000264404  0.337823       TRUE          T\n",
       "100137  rs372054324  ENSG00000264404  0.162650       TRUE          T"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load a JTI weights file to check the format\n",
    "jti_weights_path = '/data100t1/home/wanying/BioVU/202505_hypophosphatasia/data/harmonized_predixcan_db'\n",
    "weight_fn = 'JTI_Brain_Cerebellar_Hemisphere.with_snp_id.db'\n",
    "# fn = f'{jti_weights_path}/{weight_fn}'\n",
    "fn = '/data100t1/share/predixcan_models/gtex_v8/JTI/JTI_Brain_Cerebellar_Hemisphere.db'\n",
    "conn = sqlite3.connect(fn)\n",
    "# conn = sqlite3.connect('/data100t1/share/predixcan_models/gtex_v8/JTI/JTI_Whole_Blood.db')\n",
    "cur = conn.cursor()\n",
    "cmd = f\"SELECT * FROM weights\"\n",
    "df_weights = pd.read_sql_query(sql=cmd, con=conn)\n",
    "cur.close()\n",
    "\n",
    "gene = 'ENSG00000264404'\n",
    "df_weights[df_weights['gene']==gene]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d011b61-5fba-49d0-b28b-3f10ba339778",
   "metadata": {},
   "source": [
    "# 2. Prepare predixcan run in AGD 250k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5adf17e8-6b40-4784-bd8b-345acba74e5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_genotype = '/data100t1/home/wanying/BioVU/202505_hypophosphatasia/data/agd_250k_ALPL_region.vcf.gz'\n",
    "path_predixcan = '/data100t1/gapps/MetaXcan/software'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffaf651d-6f81-45dc-9aeb-08bf98fc6a83",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# conda activate py3.6\n",
    "jti_weights_path=/data100t1/home/wanying/BioVU/202505_hypophosphatasia/data/harmonized_predixcan_db\n",
    "weight_fn=JTI_Whole_Blood.with_snp_id.db \n",
    "\n",
    "# genotype_vcf=/data100t1/share/BioVU/agd_250k/vcf-converted/agd250k_chr1.primary_pass.vcf.gz\n",
    "genotype_vcf=/data100t1/home/wanying/BioVU/202505_hypophosphatasia/data/agd_250k_ALPL_region.vcf.gz\n",
    "\n",
    "# Default --model_db_snp_key is rsid. Use the new column I created\n",
    "python /data100t1/gapps/MetaXcan/software/Predict.py \\\n",
    "    --model_db_path ${jti_weights_path}/${weight_fn} \\\n",
    "    --model_db_snp_key snp_id_GRch38 \\\n",
    "    --vcf_genotypes ${genotype_vcf} \\\n",
    "    --vcf_mode genotyped \\\n",
    "    --prediction_output alpl.txt \\\n",
    "    --prediction_summary_output alpl.summary.txt \\\n",
    "    --only_entries ENSG00000162551 \\\n",
    "    --verbosity 9 \\\n",
    "    --throw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "3ba4d849-a080-4f3d-ba0e-38350a346504",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>gene</th>\n",
       "      <th>weight</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>eff_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs117815334</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>0.082700</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs2420507</td>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>-0.037631</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          rsid             gene    weight ref_allele eff_allele\n",
       "0  rs117815334  ENSG00000000419  0.082700          C          G\n",
       "1    rs2420507  ENSG00000000457 -0.037631          A          G"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load a JTI weights file to check the format\n",
    "jti_weights_path = '/data100t1/home/wanying/BioVU/202505_hypophosphatasia/data/harmonized_predixcan_db'\n",
    "weight_fn = 'JTI_Brain_Cerebellar_Hemisphere.with_snp_id.db'\n",
    "# fn = f'{jti_weights_path}/{weight_fn}'\n",
    "fn = '/data100t1/share/predixcan_models/gtex_v8/JTI/JTI_Brain_Cerebellar_Hemisphere.db'\n",
    "conn = sqlite3.connect(fn)\n",
    "# conn = sqlite3.connect('/data100t1/share/predixcan_models/gtex_v8/JTI/JTI_Whole_Blood.db')\n",
    "cur = conn.cursor()\n",
    "# cmd = f\"SELECT DISTINCT gene FROM weights\"\n",
    "cmd = f\"SELECT * FROM weights\"\n",
    "df_weights = pd.read_sql_query(sql=cmd, con=conn)\n",
    "display(df_weights.head(2))\n",
    "cur.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "2d16f191-3220-4fdf-a549-e57978a12e32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>genename</th>\n",
       "      <th>geneannot</th>\n",
       "      <th>pred.perf.R2</th>\n",
       "      <th>n.snps.in.model</th>\n",
       "      <th>pred.perf.pval</th>\n",
       "      <th>pred.perf.qval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0.046892</td>\n",
       "      <td>1</td>\n",
       "      <td>3.997963e-03</td>\n",
       "      <td>9.918923e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0.239264</td>\n",
       "      <td>9</td>\n",
       "      <td>6.486958e-12</td>\n",
       "      <td>6.120150e-11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene genename       geneannot  pred.perf.R2  n.snps.in.model  \\\n",
       "0  ENSG00000000419     DPM1  protein_coding      0.046892                1   \n",
       "1  ENSG00000000457    SCYL3  protein_coding      0.239264                9   \n",
       "\n",
       "   pred.perf.pval  pred.perf.qval  \n",
       "0    3.997963e-03    9.918923e-03  \n",
       "1    6.486958e-12    6.120150e-11  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 'weights',), ('sample_info',), ('extra',), ('construction',)\n",
    "\n",
    "# Load a JTI weights file to check the format\n",
    "jti_weights_path = '/data100t1/home/wanying/BioVU/202505_hypophosphatasia/data/harmonized_predixcan_db'\n",
    "weight_fn = 'JTI_Brain_Cerebellar_Hemisphere.with_snp_id.db'\n",
    "# fn = f'{jti_weights_path}/{weight_fn}'\n",
    "fn = '/data100t1/share/predixcan_models/gtex_v8/JTI/JTI_Brain_Cerebellar_Hemisphere.db'\n",
    "conn = sqlite3.connect(fn)\n",
    "# conn = sqlite3.connect('/data100t1/share/predixcan_models/gtex_v8/JTI/JTI_Whole_Blood.db')\n",
    "cur = conn.cursor()\n",
    "\n",
    "table = 'extra'\n",
    "cmd = f\"SELECT * FROM {table}\"\n",
    "df = pd.read_sql_query(sql=cmd, con=conn)\n",
    "df.head(2)\n",
    "\n",
    "# cur.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b02eb95-2553-45c5-ad81-7e4ea7f41b4f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95ac73e3-6966-4a68-89ad-0c2e1ea1283f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
